### ISSUE TITLE
tmux passthrough of large data sequence

### ISSUE THREAD

**Issue by wookayin** [tmux/tmux#1502]
TL;DR) tmux passthrough using <kbd>\033Ptmux;</kbd> ... <kbd>\033\\</kbd> control sequence doesn't work for a very long sequence.

## Environment

* Darwin i386 (macOS)
* tmux 2.7 (also confirmed in the latest HEAD as of 10/14)
* screen-256color (also confirmed in xterm-256color etc.)

## Description 

iTerm2 supports a file download (or image display) [protocol](https://www.iterm2.com/documentation-images.html). It works perfectly outside tmux, but **not working inside tmux** well (provided wrapped by tmux passthrough control sequence).

It works outside tmux:

![no tmux](https://user-images.githubusercontent.com/1009873/46921701-b3b40e80-cfcc-11e8-9d87-3cee631ff3bf.png)

Inside tmux, only small data sequence works. However, large files are not submitted; seems data transmission is interrupted and packets are not all sent.

![with tmux](https://user-images.githubusercontent.com/1009873/46921742-1f967700-cfcd-11e8-98e5-65d248017236.png)

For a file of 262144 bytes, the data seems to be truncated in the middle.

![image](https://user-images.githubusercontent.com/1009873/46921773-7308c500-cfcd-11e8-9c97-5c6a1a1351a5.png)

Please note that this feature is widely used for [imgcat](https://gitlab.com/gnachman/iterm2/issues/3898), etc. As a result, we cannot display a large image file inside tmux.

### How to reproduce?

To reproduce, consider a test script [`test-iterm.sh`](https://gist.github.com/wookayin/6e09c6af8612864484f791ee9297d6f7), which sends a dummy file of specified number of bytes through [iTerm2's file download protocol](https://www.iterm2.com/documentation-images.html). Note that when running under tmux, the pass-through control sequence <kbd>\033Ptmux;</kbd> ... <kbd>\033\\</kbd> is applied on OSC and ST. 

```bash
#!/bin/bash

function print_osc() {
    if [[ -n $TMUX ]] ; then
        printf "\033Ptmux;\033\033]";
    else printf "\033]"; fi
}

function print_st() {
    if [[ -n $TMUX ]] ; then
        printf "\a\033\\"
    else printf "\a"; fi
}

function send_file() {
    # sends a dummy file containing $REPEAT '.' characters
    REPEAT=${1:-1024}
    filename=$(echo -ne "test-$REPEAT.txt" | base64)
    echo "Sending a text data of $REPEAT bytes ..."

    print_osc
    printf "1337;File=name=$filename;size=$REPEAT;inline=0:"
    head -c $REPEAT < /dev/zero | tr '\0' '.' | base64
    print_st
}

send_file $1
```

## Suspects?

I have no very good idea why it is happening. Any thoughts?

* Seems related to #1388 -- input timer may just flushes data out. Not sure.
* Buffer size limit (#487), which is currently 1M. However, the data being transmitted is less than 1MB (see `test-262144.txt` example), so I doubt of it.
* #846 -- not relevant.

### What else did I try?

I tried to split the sequence into multiple chunks (e.g. use `...` <kbd>\033\\</kbd> + <kbd>\033Ptmux;</kbd> `...`), but no success.

**Comment by nicm** [comment]
What happens when you make the timer longer?

**Comment by wookayin** [comment]
Tried applying the following patch. However, I can't see any changes (still canceled at 727 bytes, and tmux just works normally).

```diff
diff --git input.c input.c
index 41cdfb7..3047f07 100644
--- input.c
+++ input.c
@@ -736,7 +736,7 @@ input_timer_callback(__unused int fd, __unused short events, void *arg)
 static void
 input_start_timer(struct input_ctx *ictx)
 {
-       struct timeval  tv = { .tv_usec = 100000 };
+       struct timeval  tv = { .tv_sec = 10, .tv_usec = 100000 };

        event_del(&ictx->timer);
        event_add(&ictx->timer, &tv);
```

**Comment by wookayin** [comment]
Another information: If I do `cat tmux-out-19781.log` (the output log) outside tmux, which would replay the data transmission, iTerm2's file transmission somehow works and the correct file gets downloaded. But it didn't work inside during the (test) tmux session.

Logfile: [tmux-out-19781.log](https://github.com/tmux/tmux/files/2476868/tmux-out-19781.log)

**Comment by wookayin** [comment]
From `tmux-server-19781.log`:

```

**Comment by wookayin** [comment]
In a quick experiment where `tty_block_maybe()` is forced to always return `0`, it successfully worked (262144 bytes). But not for 1048576 bytes (1MB) -- I believe for even larger files we are blocked by #487 (hard-coded buffer limit).

So it seems that two modification points are needed:

1. Revise `tty_block_maybe()`. Actually it compares buffer length to `TTY_BLOCK_START(tty)`, which is roughly the number of displayable characters in the tty (screen). In the pass-through mode, this shall be handled differently. Specifically it shouldn't have nothing to do with tty size. Then, how can we tell whether it is currently under pass-through mode?

1. Buffer size limit (see #487) -- This is related to the buffer size limit of dcs sequence. Is there any way to make it configurable or can we make it not limited by the buffer limit? I guess the splitting-into-chunks technique could work around this.

**Comment by nicm** [comment]
You could try to make it so that the blocking mechanism is disabled until the content of the DCS passthrough escape has been flushed to the terminal (you will need at least a flag for when it is in progress and also probably a byte count to wait until it is flushed, and will need to think about what happens if another sequence is sent while the existing one is still waiting).

But tmux is not a file transfer program and I'm not going to mess too much with the blocking code (which works extremely well for normal output) to get this to work.
